File: 07-Weighted-Graph/Code-Java/DenseWeightedGraph.java
package weightedGraph;

import java.util.Vector;

/**
 * 带权图的稠密图——邻接矩阵
 */
public class DenseWeightedGraph<Weight extends Number & Comparable> implements WeightedGraph {
    private int n;  // 节点数
    private int m;  // 边数
    private boolean directed;  // 是否为有向图
    private Edge<Weight>[][] g;  // 图的具体数据

    public DenseWeightedGraph(int n, boolean directed) {
        assert n >= 0;
        this.n = n;
        this.m = 0;  // 初始化没有任何边
        this.directed = directed;  // true 表示有向图，false 表示无向图

        // g 初始化为 n*n 的布尔矩阵, 每一个 g[i][j] 均为 null, 表示没有任何边
        this.g = new Edge[n][n];
    }

    // 返回节点个数
    @Override
    public int getV() {
        return n;
    }

    // 返回边数
    @Override
    public int getE() {
        return m;
    }

    // 向图中添加一个边
    @Override
    public void addEdge(Edge e) {
        assert e.getV() >= 0 && e.getV() < n;
        assert e.getW() >= 0 && e.getW() < n;

        if (hasEdge(e.getV(), e.getW())) return;

        g[e.getV()][e.getW()] = new Edge(e);

        // 如果为无向图
        if (e.getV() != e.getW() && !directed)
            g[e.getV()][e.getW()] = new Edge(e.getW(), e.getV(), e.getWeight());

        m ++;
    }

    // 验证图中是否有从 v 到 w 的边
    @Override
    public boolean hasEdge(int v, int w) {
        assert v >= 0 && v < n;
        assert w >= 0 && w < n;
        return g[v][w] != null;
    }

    // 返回图中一个节点的所有邻边
    @Override
    public Iterable<Edge<Weight>> adj(int v) {
        assert v >= 0 && v < n;
        Vector<Edge<Weight>> adjV = new Vector<>();
        for (int i = 0; i < n; i++)
            if (g[v][i] != null)
                adjV.add(g[v][i]);
        return adjV;
    }

    // 显示图的信息
    @Override
    public void show() {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++)
                if (g[i][j] != null)
                    System.out.print(g[i][j].getWeight() + "\t");
                else System.out.print("NULL\t");
            System.out.println();
        }
    }
}

File: codePython/07-Weighted-Graph/dense_weighted_graph.py
"""
带权图的稠密图——邻接矩阵
"""


class DenseWeightedGraph:
    def __init__(self, n, directed):
        """
        g 初始化为 n*n 的布尔矩阵, 每一个 g[i][j] 均为 None, 表示没有任何边
        :param n: 节点数
        :param directed: True 表示有向图，False 表示无向图
        """
        assert n > 0
        self.n = n
        self.m = 0  # 边数，初始化没有任何边
        self.directed = directed
        self.g = [[None for _ in range(n)] for _ in range(n)]

    def get_v(self):
        """返回节点个数"""
        return self.n

    def get_e(self):
        """返回边数"""
        return self.m

    def add_edge(self, edge):
        """向图中添加一个边"""
        assert edge.get_v() in range(self.n)
        assert edge.get_w() in range(self.n)
        if self.has_edge(edge.get_v(), edge.get_w()):
            return

        self.g[edge.get_v()][edge.get_w()] = edge

        # 如果为无向图
        if edge.get_v() != edge.get_w() and not self.directed:
            self.g[edge.get_v()][edge.get_w()] = Edge(edge.get_w(), edge.get_v(), edge.get_weight())

        self.m += 1

    def has_edge(self, v, w):
        """验证图中是否有从 v 到 w 的边"""
        assert v in range(self.n)
        assert w in range(self.n)
        return self.g[v][w]

    def adj(self, v):
        """返回图中一个节点的所有邻边"""
        return [i for i in range(self.n) if self.g[v][i]]

    def show(self):
        """显示图的信息"""
        for i in self.g:
            for j in i:
                if j:
                    print(j.get_weight(), end='\t')
                else:
                    print(j, end='\t')
            print()

